### PR TITLE
TGUI Admin DJ Panel

### PR DESCRIPTION
**Pull request by ZeWaka**
## About the PR

TGUI Admin DJ Panel

Serving as an Intermediate-Advanced tutorial

image
image

## Why's this needed? 

pretty

[FEAT][UI]

### CODE REVIEW

File: tgui/packages/tgui/format.js [hunk 1]
```diff
@@ -110,3 +110,10 @@ export const formatDb = value => {
   }
   return sign + formatted + ' dB';
 };
+
+/**
+ * Truncates a string with an ellipsis after n characters. Default is 25.
+ */
+export const truncate = (str, n = 25) => {
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/format.js:117]: What's the use-case for this over `text-overflow: ellipsis`? https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow

File: tgui/packages/tgui/interfaces/DJPanel.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+            value={frequency}
+            step={0.1}
+            minValue={-100}
+            maxValue={100}
+            format={value => {
+              return toFixed(value * 100) + '%';
+            }}
+            onDrag={(e, value) => act('set-freq', {
+              frequency: value,
+            })}
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/DJPanel.js:180]: As these are re-used for the `Knob` below, would prefer they're created outside of the `return` and then referenced both here and in the `Knob`'s props.

File: tgui/packages/tgui/interfaces/SlotMachine.js [hunk 3]
```diff
@@ -14,7 +14,7 @@ export const SlotMachine = (props, context) => {
     <Window
       title="Slot Machine"
       width={375}
-      height={175}>
+      height={190}>
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/SlotMachine.js:17]: PR scope creep is real.

File: code/modules/sound/djpanel.dm [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+			loaded_sound = foundsound
+			. = TRUE
+
+		if("set-volume")
+			var/new_volume = params["volume"]
+			if(new_volume  == "reset")
+				sound_volume = initial(sound_volume)
+				. = TRUE
+			else if(text2num(new_volume) != null)
+				sound_volume = clamp(text2num(new_volume), 0, 100)
```
Comment (mordent-goonstation) [inline code/modules/sound/djpanel.dm:77]: Good min-max check. 👍

File: tgui/packages/tgui/format.js (hunk 1 above)
Comment (ZeWaka) [reply tgui/packages/tgui/format.js:117]: that not working

Comment (mordent-goonstation) [reply tgui/packages/tgui/format.js:117]: That worries me and should be looked at. Can you create an Issue for that? On phone atm.

Comment (ZeWaka) [reply tgui/packages/tgui/format.js:117]: valid function for when you don't want to do it for overflowing imo

Comment (mordent-goonstation) [reply tgui/packages/tgui/format.js:117]: Disagree due to variable character widths, but we can argue about that another time.
